document-store/src/main/java/org/hypertrace/core/documentstore/mongo/query/parser/MongoFromTypeExpressionParser.java

```diff
@@ -42,9 +61,103 @@ public List<BasicDBObject> visit(UnnestExpression unnestExpression) {
+    Query transformedSubQuery =
+        MongoQueryTransformer.transform(subQueryJoinExpression.getSubQuery());
+    List<BasicDBObject> aggregatePipeline =
+        new ArrayList<>(mongoQueryExecutor.convertToAggregatePipeline(transformedSubQuery));
+
+    String joinedResultFieldName =
+        getJoinedResultFieldName(subQueryJoinExpression.getSubQueryAlias());
+    // Add the lookup stage to join the subquery results with the main collection
+    aggregatePipeline.add(createLookupStage(subQueryJoinExpression, joinedResultFieldName));
+
+    // Lookup Stage puts the joined results into an array field. We need to unwind that array field
+    // to get the joined results as separate documents.
+    aggregatePipeline.add(createUnwindStage(joinedResultFieldName));
+
+    // Replace root with the joined document
+    aggregatePipeline.add(createReplaceRootStage(joinedResultFieldName));
+
+    return aggregatePipeline;
+  }
+
+  private BasicDBObject createLookupStage(
+      SubQueryJoinExpression subQueryJoinExpression, String joinedResultFieldName) {
+    BasicDBObject lookupStage = new BasicDBObject();
+    BasicDBObject lookupSpec = new BasicDBObject();
+
+    lookupSpec.put("from", mongoQueryExecutor.getCollectionName());
+    lookupSpec.put("let", subQueryJoinExpression.getJoinCondition().accept(mongoLetClauseBuilder));
+    lookupSpec.put("pipeline", createLookupPipeline(subQueryJoinExpression));
+    lookupSpec.put("as", joinedResultFieldName);
+
+    lookupStage.put(LOOKUP_OPERATOR, lookupSpec);
+    return lookupStage;
+  }
+
+  private List<BasicDBObject> createLookupPipeline(SubQueryJoinExpression subQueryJoinExpression) {
+    List<BasicDBObject> pipeline = new ArrayList<>();
+    pipeline.add(createMatchStage(subQueryJoinExpression));
+    return pipeline;
+  }
+
+  private BasicDBObject createMatchStage(SubQueryJoinExpression subQueryJoinExpression) {
+    BasicDBObject matchStage = new BasicDBObject();
+    BasicDBObject expr = new BasicDBObject();
+    expr.put(EXPR_OPERATOR, getFilterClause(subQueryJoinExpression.getJoinCondition()));
+    matchStage.put(MATCH_OPERATOR, expr);
+    return matchStage;
+  }
+
+  private BasicDBObject getFilterClause(FilterTypeExpression joinCondition) {
+    final FilterTypeExpressionVisitor parser =
+        new MongoFilterTypeExpressionParser(
+            MongoRelationalFilterContext.builder()
+                .location(FilterLocation.INSIDE_EXPR)
+                .lhsParser(
+                    new MongoDollarPrefixingIdempotentParser(new MongoIdentifierExpressionParser()))
+                .rhsParser(
+                    new MongoDollarPrefixingIdempotentParser(
+                        new MongoAliasedIdentifierExpressionParser()))
+                .build());
+    final Map<String, Object> filter = joinCondition.accept(parser);
+    return new BasicDBObject(filter);
+  }
+
+  private String getJoinedResultFieldName(String subQueryAlias) {
+    return "__joined_result_with_" + subQueryAlias;
```

Comment:
Could you please explain what this is for?

Reply:
This is the field name in which the joined results will be put after the lookup stage. This field name can be used by subsequent stages in the aggregate pipeline to do operations on the joined result.
^Also added it as a comment